src/query/storages/stage/src/read/row_based/processors/reader.rs

```diff
@@ -48,30 +46,31 @@ impl BytesReader {
     pub fn try_create(
         table_ctx: Arc<dyn TableContext>,
         op: Operator,
-        read_batch_size: usize,
+        _read_batch_size: usize,
         prefetch_num: usize,
     ) -> Result<Self> {
         Ok(Self {
             table_ctx,
             op,
-            read_batch_size,
+            // TODO(xuanwo): bring batch size back when we figure how to play with range based read.
+            read_batch_size: 4 * 1024 * 1024,
             file_state: None,
             prefetch_num,
         })
     }
 
     pub async fn read_batch(&mut self) -> Result<DataBlock> {
         if let Some(state) = &mut self.file_state {
-            let end = min(self.read_batch_size + state.offset, state.file.size);
-            let mut buffer = vec![0u8; end - state.offset];
-            let n = read_full(&mut state.reader, &mut buffer[..]).await?;
-            if n == 0 {
+            let end = min(self.read_batch_size + state.offset, state.file.size) as u64;
+
+            let buffer = state.reader.read(state.offset as u64..end).await?.to_vec();
```

Comment:
I used to read like this mistakingly.
FYI, change bache size to 16M is not helpful.